ts/Extensions/Breadcrumbs.ts

```diff
@@ -0,0 +1,898 @@
+                        fill: 'white',
+                        stroke: 'black',
+                        'stroke-width': 1,
+                        rotation: serparatorRotation,
+                        rotationOriginX: posX,
+                        rotationOriginY: posY
+                    })
+                    .add(breadcrumbs.breadcrumbsGroup)
+                    .addClass('highcharts-breadcrumbs-separator'),
+                separatorBBox = separator.getBBox();
+
+            separator.translate(separatorBBox.width, separatorBBox.height / 2);
+
+            return separator;
+        }
+    }
+
+    /**
+    * Update.
+    * @function Highcharts.Breadcrumbs#update
+    *
+    * @requires  modules/breadcrumbs
+    *
+    * @param {Highcharts.Breadcrumbs} this
+    *        Breadcrumbs class.
+    * @param {Highcharts.BreadcrumbsOptions} userOptions
+    *        Breadcrumbs class.
+    * @param {boolean} redrawFlag
+    *        Redraw flag
+    */
+    public update(this: Breadcrumbs, userOptions: Highcharts.Options, redrawFlag: boolean = true): void {
+        merge(true, this.chart.options.breadcrumbs, userOptions);
+
+        if (redrawFlag) {
+            this.destroy(-1);
+            this.redraw();
+        }
+    }
+}
+
+/* eslint-disable no-invalid-this */
+
+if (!H.Breadcrumbs) {
+    H.Breadcrumbs = Breadcrumbs as any;
+
+    addEvent(Chart, 'getMargins', function (e: Chart): void {
+        const chart = this,
+            breadcrumbsOptions = chart.options.breadcrumbs,
+            defaultBreadcrumheight = 35, // TO DO: calculate that height
+            extraMargin = chart.rangeSelector ? chart.rangeSelector.getHeight() : defaultBreadcrumheight;
+
+        if (
+            breadcrumbsOptions &&
+            breadcrumbsOptions.enabled &&
+            !breadcrumbsOptions.floating &&
+            chart.breadcrumbs &&
+            chart.breadcrumbs.breadcrumbsList &&
+            chart.breadcrumbs.breadcrumbsList.length
+        ) {
+            if (breadcrumbsOptions.position && breadcrumbsOptions.position.verticalAlign === 'top' && extraMargin) {
+                chart.plotTop += extraMargin;
+            }
+            if (breadcrumbsOptions.position && breadcrumbsOptions.position.verticalAlign === 'bottom' && extraMargin) {
+                (chart.marginBottom as number) += extraMargin;
+            }
+        }
+    });
+
+    addEvent(Chart, 'redraw', function (e: Chart): void {
+        this.breadcrumbs?.redraw();
+    });
+
+    addEvent(Chart, 'afterDrilldown', function (e: Chart): void {
+        const chart = this,
+            breadcrumbs = chart.breadcrumbs as Highcharts.Breadcrumbs,
+            breadcrumbsOptions = chart.options.breadcrumbs;
+
+        if (breadcrumbsOptions && breadcrumbsOptions.enabled) {
+
+            if (!breadcrumbs) {
+                chart.breadcrumbs = new Breadcrumbs(chart as Chart, chart.options as Highcharts.Options);
+            }
+            chart.redraw();
```

Comment:
This is an interesting case:
- drilldown call `chart.redraw()`, then fires `afterDrilldown` event
- breadcrumbs add `afterDrilldown` and fires `chart.redraw()`

I think we can somehow improve this, in future similar cases can happen too. I see two possible solutions:

- create `afterDrilldownBeforeRedraw` event and place it before `chart.redraw()`
- move `fireEvent(this, 'afterDrilldown');` before `chart.redraw()`

The first one pollutes events scope. The second one might have unexpected results and bugs. The current solution (multiple redraws) is from the performance perspective is a bad idea.

Any other ideas how to solve it? What do you think @TorsteinHonsi ?

Reply:
Yes, multiple redraws should be avoided. Other than that I don't have full overview over this, but I prefer alternative two instead of adding specialized event calls.

Reply:
This is a more general issue, the second solution does not resolve this issue. The same problem we have with RangeSelector, Navigator, etc components that need to resize the plotting area. Some components (title/subtitle and legend) already do this in `chart.redraw()`. We will try to do this:
- create method in `chart` that will internally call `chart.layoutTitles()` and `legend.render()`
- method will fire event, allowing other chart components to reserve some space in the chart
- RangeSelector may now remove workaround, implement new event solution and fix a related bug - that will additionally improve performance (will be separate PR)
- Navigator may now easily fix related bugs (where Navigator width does not match the plotting area) -  separate bugfix/PR
- StockTools - no one reported a bug, but I will check this